src/lsp/cobol_parser/grammar.mly

```diff
@@ -3998,17 +3998,38 @@ let start_position ==
 (* STOP STATEMENT *)
 
 %public let unconditional_action := ~ = stop_statement; <Stop>
-let stop_statement [@context stop_stmt] :=
-  | STOP; RUN; so = ro(with_status); { StopRun so }       (* status: +COB2002 *)
-  | STOP; l = literal;               { StopLiteral l }    (* ~COB85, -COB2002 *)
+let stop_statement :=
+ | STOP; ~ = stop_body; < >
+let stop_body [@context stop_stmt] :=
+  | ~ = o(stop_with_arg); <StopArg> (* RM/COBOL extension *)
+  | RUN; ~ = o(stop_run_returning_body); <StopRun>
+  | ERROR; { StopError }                         (* GCOS *)
+
+let stop_run_returning_body :=
+  | ~ = scalar; <StopReturnScalar> 
+  | or_(GIVING, RETURNING); o(ADDRESS; OF); ~ = qualident; <StopReturnAddress>
+  | or_(GIVING, RETURNING); v = integer;
+    { StopReturnInt { value = Integer v;
+                      size = None } }
+  | or_(GIVING, RETURNING); v = integer; s = pf(SIZE; IS, integer);
+    { StopReturnInt { value = Integer v;
+                      size = Some (Integer s) } }
+  | ~ = with_status; <StopReturnStatus>
```

Comment:
Are all those in the standard ? Or are they mostly extensions ?

Reply:
There are a lot of extensions (common, RM/COBOL and GCOS according to GnuCOBOL). I did not find information on `| stop thread_reference_optional` and I did not implement it. I think I'll do it in this PR.